lineax/_solver/lsmr.py

```diff
@@ -0,0 +1,583 @@
+"""Implementation adapted from SciPy, with BSD license:
+
+Copyright (c) 2001-2002 Enthought, Inc. 2003-2024, SciPy Developers.
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+   notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above
+   copyright notice, this list of conditions and the following
+   disclaimer in the documentation and/or other materials provided
+   with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+   contributors may be used to endorse or promote products derived
+   from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+
+
+from collections.abc import Callable
+from typing import Any, Optional
+from typing_extensions import TypeAlias
+
+import jax.lax as lax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+from equinox.internal import ω
+from jaxtyping import Array, PyTree
+
+from .._norm import two_norm
+from .._operator import AbstractLinearOperator, conj
+from .._solution import RESULTS
+from .._solve import AbstractLinearSolver
+
+
+_LSMRState: TypeAlias = AbstractLinearOperator
+
+
+class LSMR(AbstractLinearSolver[_LSMRState], strict=True):
+    """LSMR solver for linear systems.
+
+    This solver can handle any operator, even nonsquare or singular ones. In these
+    cases it will return the pseudoinverse solution to the linear system.
+
+    Similar to `scipy.sparse.linalg.lsmr`.
+
+    This supports the following `options` (as passed to
+    `lx.linear_solve(..., options=...)`).
+
+    - `y0`: The initial estimate of the solution to the linear system. Defaults to all
+        zeros.
+    - `damp`: Damping factor for regularized least-squares. LSMR solves the regularized
+        least-squares problem:
+
+            min ||A y - b||_2 + damp*||y||_2
+
+        where damp is a scalar. If damp is None or 0, the system is solved without
+        regularization. Default is 0.
+
+    """
+
+    atol: float
+    btol: float
```

Comment:
The convention we've been using through the Equinox scientific ecosystem is to use `atol` (absolute tolerance) and `rtol` (relative tolerance). Different parts of the numerical community keep cooking up different kinds of convergence criteria for different linear/nonlinear/etc. problems, so we're attempting to impose a little order :D

Would it be possible to switch this over in the same way?

FWIW you should be able to arrange for a solver to take exactly `n` steps by setting `atol=rtol=0` and `max_steps=n`, `lx.linear_solve(..., throw=False)`. This makes it possible to compare numerical results across implementations, if that is desired.

Reply:
These are a bit confusingly named, since they are both actually relative tolerances, but relative to `a` (the linear operator) or `b` (the rhs vector). I think we could probably change it, since later on they both get reduced to a single relative tolerance:

```python
            rtol = self.btol + self.atol * normA * normx / normb
```

I'd have to think about an absolute tolerance, since it's a bit weird for least squares where the residual is never expected to be zero.